### PR TITLE
Add of toolbar with interactives save button, pans, home button, zoom…

### PR DESCRIPTION
**Pull request by jujusb**
Add of toolbar with interactives save button with name by default already given, pans, home button, zoom, expand the figure size for the matplotlib plot

### CODE REVIEW

File: vis/xAPISG-CompletablesScores.ipynb [hunk 1]
```diff
@@ -73,19 +76,36 @@
...
+    "            name=filename+\"_completable_scores\"\n",
+    "            if nbgraphshown>1:\n",
+    "                name+=\"_\"+str(n+1)+\"_\"+str(nbgraphshown)\n",
+    "            fig = plt.figure(num=name, constrained_layout=True)\n",
+    "            fig.canvas.header_visible = False\n",
+    "            fig.canvas.layout.min_height = '400px'\n",
+    "            figCompletableScore.append(fig)\n",
+    "            if xaxisPerCompletable:\n",
+    "                players_scores[n] = reorganiseDictBySubKeys(players_scores[n])\n",
+    "            if len(completableSelected)>0 and len(playersSelected)>0::\n",
```
Comment (crisal24) [inline vis/xAPISG-CompletablesScores.ipynb]: SyntaxError ::

File: vis/xAPISG-CompletablesScores.ipynb [hunk 2]
```diff
@@ -73,19 +76,36 @@
...
-    "        else:\n",
-    "            noDataToFillVis(100)\n",
-    "        plt.xlabel(\"Completable\")\n",
-    "        plt.ylabel(\"Score\")\n",
-    "        plt.xticks(rotation=90)\n",
-    "        plt.title(\"Scores of players per completable\")\n",
-    "        plt.show()"
+    "        accordion[n]=widgets.Output()\n",
+    "        with accordion[n]:\n",
+    "            n_groups = len(players_scores[n])\n",
```
Comment (crisal24) [inline vis/xAPISG-CompletablesScores.ipynb]: n_groups never used

File: vis/xAPISG-CompletablesProgress.ipynb [hunk 3]
```diff
@@ -74,21 +77,38 @@
...
+    "        accordion[n]=widgets.Output()\n",
+    "        with accordion[n]:\n",
+    "            name=os.path.splitext(filename)[0]+\"_completable_progress\"\n",
+    "            if nbgraphshown>1:\n",
+    "                name+=\"_\"+str(n+1)+\"_\"+str(nbgraphshown)\n",
+    "            fig = plt.figure(num=name, constrained_layout=True)\n",
+    "            fig.canvas.header_visible = False\n",
+    "            fig.canvas.layout.min_height = '400px'\n",
+    "            figCompletableProgress.append(fig)\n",
+    "            n_groups = len(players_progress[n])\n",
```
Comment (crisal24) [inline vis/xAPISG-CompletablesProgress.ipynb]: n_groups never used

File: vis/xAPISG-CompletablesTimes.ipynb [hunk 4]
```diff
@@ -101,38 +104,55 @@
     "        min_times=separateListToSubListEachXvalues(min_times, xaxis)\n",
     "        N=xaxis\n",
     "        nbgraphshown=len(max_times)\n",
+    "    accordion={}\n",
     "    # Bar chart showing for each completable the maximum, minimum and mean time of completion by players.\n",
     "    for n in range(nbgraphshown):\n",
-    "        if n_groups>0 :\n",
```
Comment (crisal24) [inline vis/xAPISG-CompletablesTimes.ipynb:111]: n_groups assignment (L79) now never used

File: xAPISGProcessor.ipynb [hunk 5]
```diff
@@ -74,10 +74,25 @@
...
    ]
   },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "outputs": [],
+   "source": [
+    "! jupyter nbextension enable --py --sys-prefix widgetsnbextension\n",
+    "! jupyter nbextension install --py --symlink --sys-prefix ipympl\n",
+    "! jupyter nbextension enable ipympl --py --sys-prefix"
```
Comment (crisal24) [inline xAPISGProcessor.ipynb]: Is there another way to do this? Seems like too low level code to have it in the main notebook.

**Review by crisal24 — CHANGES_REQUESTED** [review]
Great work! After fixing a syntax error (see comment) and installing the new ipympl, the new toolbar was correctly displayed with great features to better visualize the data and save the plots. This could greatly improve the user experience!

Some issues found: 
- axis labels and titles seems not to be displayed now
- some top values in bar charts appear outside of the scope of the visualization

General comment about the code: as the code of the visualizations has increased with the new features, it could be helpful to add more comments in the code to help readers understand what is been done at each step. Consider revise the current code to improve documentation and comments as a next step (when funcionality is completed).

File: vis/xAPISG-AccessedAccessible.ipynb [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+    "collapsed": false,
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "**** vis_accessibles_function_time function** * display a bubble chart showing a bubble in function of the average of players has accessed this accessible in a certain period of time."
```
Comment (crisal24) [inline vis/xAPISG-AccessedAccessible.ipynb]: fix markdown **

File: vis/xAPISG-AlternativesSelectedQuestion.ipynb [hunk 7]
```diff
@@ -109,15 +143,40 @@
...
+    "                plt.yticks(np.arange(0, total_max[0], 1))\n",
+    "                plt.ylim(0,total_max[0])\n",
+    "                #add axes labels and title\n",
+    "                plt.xlabel(\"Responses selected\")\n",
+    "                plt.ylabel(\"Number of responses\")\n",
+    "                plt.show()\n",
+    "    else:\n",
+    "        #create output for this graph\n",
+    "        accordion[0]=widgets.Output()\n",
+    "        #TODO only display the one selected in the accordion\n",
```
Comment (crisal24) [inline vis/xAPISG-AlternativesSelectedQuestion.ipynb]: Missing TODO?

File: vis/xAPISG-GamesStartedCompleted.ipynb [hunk 8]
```diff
@@ -29,23 +28,68 @@
     }
    }
   },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "outputs": [
```
Comment (crisal24) [inline vis/xAPISG-GamesStartedCompleted.ipynb]: Clear NameError output

File: vis/xAPISG-ItemsInteracted.ipynb [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    "    #get all players and items selected\n",
+    "    playersSelected=checkboxesPlayersSelected.get_selected_options()\n",
+    "    items=getSublistSelectedFromUnionListSelected(\n",
+    "        item_interaction_list,\n",
+    "        checkboxesItemInteractedSelected.get_selected_options()\n",
+    "    )\n",
+    "    nbitems=constructMultiTab(players_info, playersSelected, items, \"interactions\", True)\n",
+    "    # Multiple bar chart showing the items selected for each personn.\n",
+    "    if len(nbitems)>0:\n",
+    "        #TODO only display the one selected in the accordion\n",
```
Comment (crisal24) [inline vis/xAPISG-ItemsInteracted.ipynb]: Missing TODO?

File: vis/xAPISG-MenusSelected.ipynb [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    "                        if not selection in nbmenus[menu][player].keys():\n",
+    "                            nbmenus[menu][player][selection]=0\n",
+    "                        nbmenus[menu][player][selection]+=len(players_info[player][\"selected_menus\"][menu][selection])\n",
+    "    \n",
+    "    # Multiple bar chart showing the response selected in a menu for each personn.\n",
+    "    alls_menus_vis={}\n",
+    "    if len(playersSelected)>0 and len(menus)>0:\n",
+    "        for menu in menus:\n",
+    "            alls_menus_vis[menu]=widgets.Output()\n",
+    "            #TODO only display the one selected in the accordion\n",
```
Comment (crisal24) [inline vis/xAPISG-MenusSelected.ipynb]: Missing TODO?

**Review by crisal24 — CHANGES_REQUESTED** [review]
Great work! Clear improvement in the visualizations & in the code with more comments to follow the process.

Ongoing issues: 
- No labels are displayed in the visualizations axis
- When updating the configuration in some visualizations (e.g. order), instead of updating the current visualization, the old is maintained, and the new appears below, which is a bit confusing.

File: xAPISGProcessor.ipynb [hunk 11]
```diff
@@ -70,12 +58,11 @@
...
     "from collections import Counter\n",
     "import pandas as pd\n",
-    "from ipywidgets import interact, interactive, HBox, Layout,VBox\n",
+    "from ipywidgets import interact, interactive, fixed, HBox, Layout,VBox\n",
     "import ipywidgets as widgets\n",
     "from IPython.display import display, clear_output\n",
     "import pprint\n",
-    "plt.style.use('dark_background')\n",
-    "plt.rcParams['figure.figsize']=[10, 8]"
+    "plt.style.use('dark_background')"
```
Comment (crisal24) [inline xAPISGProcessor.ipynb]: Could use the style "classic" as default? Or allow selection of style.

File: vis/clearFigMatplotlib.ipynb [hunk 12]
```diff
@@ -0,0 +1,306 @@
+{
```
Comment (crisal24) [inline vis/clearFigMatplotlib.ipynb]: Could this file be in helpersFunction folder?

File: vis/helpersFunctions/HeatMap.ipynb [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+  {
+   "cell_type": "markdown",
+   "source": [
+    "***heatMap* function** to display the heatMap\n",
+    "\n",
+    "Input :\n",
+    "* multiList: the list with the values for the HeatMap\n",
+    "* ticksXlabel : labels for the ticks of X axe of the heatmap\n",
+    "* ticksYlabel : labels for the ticks of Y axe of the heatmap\n",
+    "\n",
```
Comment (crisal24) [inline vis/helpersFunctions/HeatMap.ipynb:69]: Add inputs: figTab and figtitle

File: vis/helpersFunctions/MultiBarSeparated.ipynb [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+   "source": [
+    "***multiBarSeparated* function** to display the multiBar chart\n",
+    "\n",
+    "Input :\n",
+    "* dictionary: the list with all the values\n",
+    "* order : selection to order values alphabeticly or per values\n",
+    "* title : title of the multibar graph\n",
+    "* xlabel : labels for the X axe\n",
+    "* ylabel : labels for the Y axe\n",
+    "\n",
```
Comment (crisal24) [inline vis/helpersFunctions/MultiBarSeparated.ipynb:117]: add filetitle input

File: vis/xAPISG-CompletablesScores.ipynb [hunk 15]
```diff
@@ -7,7 +7,7 @@
    "source": [
     "%run vis/xAPISG-noDataToFillVisualization.ipynb # notebook to create the visualisation with a message NoDataToFill\n",
     "%run vis/helpersFunctions/barCharts.ipynb\n",
-    "%run vis/helpersFunctions/OrderGraphValuesAndSelectDataByXaxis.ipynb"
+    "%run vis/helpersFunctions/OrderGraphValuesAndSelectDataByXaxis.ipyn"
```
Comment (crisal24) [inline vis/xAPISG-CompletablesScores.ipynb]: ipynb?

**Review by crisal24 — CHANGES_REQUESTED** [review]
Good work! Some minor changes.

File: xAPISGProcessor.ipynb [hunk 16]
```diff
@@ -87,82 +74,288 @@
...
+    "global timeformats\n",
     "timeformats=['%Y-%m-%dT%H:%M:%SZ','%Y-%m-%dT%H:%M:%S.%fZ'] #array of time format\n",
     "\n",
     "%run fileBrowserAndUploadButtonToLoadProcessStatements.ipynb\n",
     "%run globalsSelectors.ipynb\n",
+    "%run vis/helpersFunctions/clearFigMatplotlib.ipynb\n",
+    "\n",
+    "tabsTitle=[\"Start/Complete games\",\"Videos\",\"Progress\",\"Completable\",\"Alternative\",\"Items interactions\",\"Accessibles\",\"Menus\"]\n",
+    "\n",
+    "def on_selected_tab_index_change(change):\n",
```
Comment (crisal24) [inline xAPISGProcessor.ipynb]: Could this code be hidden in another notebook?

File: globalsSelectors.ipynb [hunk 17]
```diff
@@ -38,11 +14,17 @@
    "execution_count": null,
    "outputs": [],
    "source": [
-    "#number of data per visualisation with x axis\n",
-    "global xaxis\n",
-    "xaxis=widgets.Dropdown(options=[('Alls datas',0),('5 per visualisation',5),('25 per visualisation',25),('50 per visualisation',50),('10 first',-1),('10 last',-2)],\n",
-    "                                description='x axe',\n",
-    "                                disabled=False)"
+    "global style\n",
```
Comment (crisal24) [inline globalsSelectors.ipynb:17]: Great addition! Maybe the selection is a bit complex now, could the dropdown be limited to a selection of all the available styles? (e.g. up to 10 which are different enough between them)

**Review by crisal24 — CHANGES_REQUESTED** [review]
Good additions! See minor changes in the comments.

Also, revise the size of the output view, which gets smaller when changing between tabs.

**Comment by jujusb** [comment]
> Good additions! See minor changes in the comments.
> 
> Also, revise the size of the output view, which gets smaller when changing between tabs.

The size of the output can be corrected but you need to install nodejs in your computer to execute the line that correct that.

**Review by crisal24 — APPROVED** [review]
Great work! Display of visualizations has clearly improved! Ready to merge

**Pull request merged**